src/folding/nova/decider.rs

```diff
@@ -0,0 +1,647 @@
+/// This file implements the onchain (Ethereum's EVM) decider circuit. For non-ethereum use cases,
+/// other more efficient approaches can be used.
+use ark_crypto_primitives::crh::poseidon::constraints::CRHParametersVar;
+use ark_crypto_primitives::sponge::{poseidon::PoseidonConfig, Absorb};
+use ark_ec::{CurveGroup, Group};
+use ark_ff::PrimeField;
+use ark_r1cs_std::{
+    alloc::{AllocVar, AllocationMode},
+    boolean::Boolean,
+    eq::EqGadget,
+    fields::{fp::FpVar, nonnative::NonNativeFieldVar, FieldVar},
+    groups::GroupOpsBounds,
+    prelude::CurveVar,
+    ToConstraintFieldGadget,
+};
+use ark_relations::r1cs::{ConstraintSynthesizer, ConstraintSystemRef, Namespace, SynthesisError};
+use ark_std::{One, Zero};
+use core::{borrow::Borrow, marker::PhantomData};
+
+use crate::ccs::r1cs::R1CS;
+use crate::folding::nova::{
+    circuits::{CommittedInstanceVar, FCircuit, CF1, CF2},
+    ivc::IVC,
+    CommittedInstance, Witness,
+};
+use crate::pedersen::Params as PedersenParams;
+use crate::utils::gadgets::{
+    hadamard, mat_vec_mul_sparse, vec_add, vec_scalar_mul, SparseMatrixVar,
+};
+
+#[cfg(not(test))]
+use crate::folding::nova::cyclefold::{CycleFoldCommittedInstanceVar, CF_IO_LEN};
+#[cfg(not(test))]
+use crate::pedersen::PedersenGadget;
+#[cfg(not(test))]
+use ark_r1cs_std::ToBitsGadget;
+
+#[derive(Debug, Clone)]
+pub struct RelaxedR1CSGadget<F: PrimeField, CF: PrimeField, FV: FieldVar<F, CF>> {
+    _f: PhantomData<F>,
+    _cf: PhantomData<CF>,
+    _fv: PhantomData<FV>,
+}
+impl<F: PrimeField, CF: PrimeField, FV: FieldVar<F, CF>> RelaxedR1CSGadget<F, CF, FV> {
+    /// performs the RelaxedR1CS check (Az∘Bz==uCz+E)
```

Comment:
hmm but we need to ensure that $Az \circ Bz$ is correct, if we input it already calculated we're not checking it with constraints and it could be manipulated, no?

Reply:
Also the hadamard product is for vectors, where $A, B$ are matrices, so $A \circ B$ is not possible.

Maybe we could work with a variable vector $x=(x_1, x_2, ..., x_n)$ and compute $A \cdot x$ and $B \cdot x$ (where $\cdot$ is the matrix-vector-product), so then we can compute the hadamard product of $Ax \circ Bx$ and input that into the circuit.

For example, let

$$
A=\begin{pmatrix}
0 & 1 & 0 & 0 & 0 & 0\\
0 & 0 & 0 & 1 & 0 & 0\\
0 & 1 & 0 & 0 & 1 & 0\\
5 & 0 & 0 & 0 & 0 & 1\\
\end{pmatrix}
$$

$$B=\begin{pmatrix}
0 & 1 & 0 & 0 & 0 & 0\\
0 & 1 & 0 & 0 & 0 & 0\\
1 & 0 & 0 & 0 & 0 & 0\\
1 & 0 & 0 & 0 & 0 & 0\\
\end{pmatrix}$$

and

$$x=\begin{pmatrix}
x_1\\
x_2\\
x_3\\
x_4\\
x_5\\
x_6\\
\end{pmatrix}$$

Then,

$$Az =
\begin{pmatrix}
0 & 1 & 0 & 0 & 0 & 0\\
0 & 0 & 0 & 1 & 0 & 0\\
0 & 1 & 0 & 0 & 1 & 0\\
5 & 0 & 0 & 0 & 0 & 1\\
\end{pmatrix}
\cdot
\begin{pmatrix}
x_1\\
x_2\\
x_3\\
x_4\\
x_5\\
x_6\\
\end{pmatrix}=
\begin{pmatrix}x_2 & x_4 & x_2+x_5 & 5x_1+x_6\end{pmatrix}
$$

and

$$Bz =
\begin{pmatrix}
0 & 1 & 0 & 0 & 0 & 0\\
0 & 1 & 0 & 0 & 0 & 0\\
1 & 0 & 0 & 0 & 0 & 0\\
1 & 0 & 0 & 0 & 0 & 0\\
\end{pmatrix}
\cdot
\begin{pmatrix}
x_1\\
x_2\\
x_3\\
x_4\\
x_5\\
x_6\\
\end{pmatrix}=
\begin{pmatrix}x_2 & x_2 & x_1 & x_1\end{pmatrix}
$$

and then

$$Az \circ Bz =
\begin{pmatrix}x_2^2 & x_2 x_4 & x_1 x_2 + x_1 x_5 & 5x_1^2 + x_1 x_6\end{pmatrix}
$$

Maybe we could input $Az \circ Bz$ as kind of hardcoded constraints in the circuit, so the prover can not cheat with it.

Reply:
hmm and then the same could be done for $C \cdot z$ inputing $C \cdot x$ where $x$ is the variables vector. And then the same could be done directly for the combination of $Ax \circ Bx - Cx$, getting just a single vector of combinations of $x_i$.
So in this way the whole check which originally involves matrix multiplications, gets reduced to a vector of $x_i$ which contains the respective operations already there. hmm

Reply:
Since we use sparse matrix representations, I wonder if by using this 'precomputed' $x_i$ vector we would end up doing almost the same amount of operations that we would do doing the sparse matrix operations.

Reply:
That's a fair point though.. I did not count with the fact of the sparsity. 

Still my proposal might not be feasible, we should maybe check with some people!

Reply:
I'm trying to compute $A \circ B$ and then multiplying it by $z$ to see if it leads to the same vector as if we first compute $Az$ and $Bz$ and then compute their hadamard product, ie. $(A \circ B) z$ vs $Az \circ Bz$.
*(continuing with the values from the previous example)*
$$A \circ B =
\begin{pmatrix}
0 & 1 & 0 & 0 & 0 & 0\\
0 & 0 & 0 & 1 & 0 & 0\\
0 & 1 & 0 & 0 & 1 & 0\\
5 & 0 & 0 & 0 & 0 & 1\\
\end{pmatrix}
\circ
\begin{pmatrix}
0 & 1 & 0 & 0 & 0 & 0\\
0 & 1 & 0 & 0 & 0 & 0\\
1 & 0 & 0 & 0 & 0 & 0\\
1 & 0 & 0 & 0 & 0 & 0\\
\end{pmatrix}$$

$$=\begin{pmatrix}
0 & 1 & 0 & 0 & 0 & 0\\
0 & 0 & 0 & 0 & 0 & 0\\
0 & 0 & 0 & 0 & 0 & 0\\
5 & 0 & 0 & 0 & 0 & 0\\
\end{pmatrix}$$

which when multiplied by $z = (1, 3, 35, 9, 27, 30)$ (a valid witness for this R1CS), yields

$$(A \circ B) z =
\begin{pmatrix}
0 & 1 & 0 & 0 & 0 & 0\\
0 & 0 & 0 & 0 & 0 & 0\\
0 & 0 & 0 & 0 & 0 & 0\\
5 & 0 & 0 & 0 & 0 & 0\\
\end{pmatrix}
\begin{pmatrix}
1\\
3\\
35\\
9\\
27\\
30\\
\end{pmatrix}
=\begin{pmatrix}3 & 0 & 0 & 5\end{pmatrix}
$$

But if we compute $Az \circ Bz$:

$$Az \circ Bz =
\begin{pmatrix}
0 & 1 & 0 & 0 & 0 & 0\\
0 & 0 & 0 & 1 & 0 & 0\\
0 & 1 & 0 & 0 & 1 & 0\\
5 & 0 & 0 & 0 & 0 & 1\\
\end{pmatrix}
\begin{pmatrix}
1\\
3\\
35\\
9\\
27\\
30\\
\end{pmatrix}
\circ
\begin{pmatrix}
0 & 1 & 0 & 0 & 0 & 0\\
0 & 1 & 0 & 0 & 0 & 0\\
1 & 0 & 0 & 0 & 0 & 0\\
1 & 0 & 0 & 0 & 0 & 0\\
\end{pmatrix}
\begin{pmatrix}
1\\
3\\
35\\
9\\
27\\
30\\
\end{pmatrix}
$$

$$=\begin{pmatrix}
3 & 9 & 30 & 35
\end{pmatrix}
\circ
\begin{pmatrix}
3 & 3 & 1 & 1
\end{pmatrix}
= \begin{pmatrix}
9 & 27 & 30 & 35
\end{pmatrix}
$$

Reply:
We've commented it in a call, and for the moment we're leaving the code as it is and we will explore it in the future to see if we can optimize it and bring the number of constraints down here.